### ISSUE TITLE
Attribute (region) does not pass the type constraint because: Validation failed for 'Str' with value undef

### ISSUE THREAD

**Issue by ssimms** [rustyconover/net-amazon-s3#50]
In versions 0.83 through 0.85 and master, with a bucket in us-east-1, I'm getting the above error with the following code:

```
#!/usr/bin/perl

use Net::Amazon::S3;
my $s3 = Net::Amazon::S3->new({
    aws_access_key_id => $key,
    aws_secret_access_key => $secret,
});
my $client = Net::Amazon::S3::Client->new(s3 => $s3);
my $bucket = $client->bucket(name => 'mybucket');

# dies
print $bucket->region();

# dies
print $bucket->location_constraint();

# dies
my $object = $bucket->object(key => 'path/to/object')->get();
```

It works if I use signature V2 instead of V4.  It works in versions 0.80 through 0.82, and is probably related to the fix for issue #29.

Digging through the code, it seems like there might be a problem in the way the V4 signature determines the region:

- `Net::Amazon::S3::Signature::V4->_sign` calls `_bucket_region`
- `_bucket_region` calls `http_request->region`
- `Net::Amazon::S3::HTTPRequest->region` defaults to `bucket->region`
- `Net::Amazon::S3::Client::Bucket->region` defaults to `location_constraint`
- `location_constraint` makes an HTTP request that needs to be signed

Can you reproduce this?

**Comment by branislav-zahradnik-gdc** [comment]
Hi @ssimms 

thanks for report.

`Net::Amazon::S3::HTTPRequest->region` defaults to `Net::Amazon::S3::Bucket->region`, see [`around BUILDARGS`](https://github.com/rustyconover/net-amazon-s3/blob/master/lib/Net/Amazon/S3/Role/Bucket.pm#L7)

That one defaults to [`_head_region`](https://github.com/rustyconover/net-amazon-s3/blob/master/lib/Net/Amazon/S3/Bucket.pm#L667)

I already stumbled upon this issue this week but I'm not able to reliably reproduce it.
First fix will be to dump this response (to collect better data) and retry operation few times, see #51

**Comment by ssimms** [comment]
Thanks for adding the logging -- it made the problem much clearer:

```
500 Can't connect to my.bucket.name.s3.amazonaws.com:443 (hostname verification failed)
```

It fails when there's a period in the bucket name (e.g. bucket.mydomain.com), and succeeds when there isn't one (e.g testbucket20190412).

**Comment by ranguard** [comment]
From the docs:

https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

When you use virtual hosted–style buckets with Secure Sockets Layer (SSL), the SSL wildcard certificate only matches buckets that don't contain periods. To work around this, use HTTP or write your own certificate verification logic. We recommend that you do not use periods (".") in bucket names when using virtual hosted–style buckets.


My guess is this is what you are hitting

**Comment by ssimms** [comment]
Agreed.  I just tried it with `secure => 0`, which worked for testing purposes.

Setting `use_virtual_host => 1` doesn't change the behavior.  Neither does setting `region` explicitly when instantiating the bucket -- as far as I can tell, only the bucket name is passed to the signature code, which recreates the object without the explicit region.

Signature V2 still works, but is being discontinued in a couple of months, so it won't be a long-term workaround:

https://docs.aws.amazon.com/AmazonS3/latest/dev/auth-request-sig-v2.html

> After June 24, 2019, Amazon S3 will only accept API requests that are signed using Signature Version 4.

**Comment by ssimms** [comment]
Setting `use_virtual_host => 0` solves the problem.

**Comment by rocketgithub** [comment]
0.86 works like a champ for me! It fixed the region problem. Thanks a lot for your work.

**Comment by branislav-zahradnik-gdc** [comment]
Thank you guys for identifying issue, I'm currently trying to write reasonable integration test suite, bucket with dot + SSL should be there as well

**Comment by k-asm** [comment]
Hello, @ssimms.
I encountered the same problem, but solved it with 0.85.

Check whether `LWP::Protocol::https` is installed in your environment.

This HEAD request returns 501 response when `LWP::Protocol::https` is not installed, so `region` turns to be undef.
https://github.com/rustyconover/net-amazon-s3/blob/master/lib/Net/Amazon/S3/Bucket.pm#L699

**Comment by ssimms** [comment]
@k-asm,

I've had that problem in the past, so it was one of the first things I checked before finding out that the problem was related to the bucket's name.  :-)

```
$ perl -MLWP::Protocol::https -E 'say $LWP::Protocol::https::VERSION'
6.07
```

**Comment by Vultour** [comment]
This still breaks non-AWS S3 services because the HEAD request can fail. There should be a way to choose between `_head_region` or a custom value.

**Comment by happy-barney** [comment]
@Vultour I started to work (well, it's a plan for upcoming weekend ...) on extracting such information / behaviour into `Vendor` context. May I ask you what kind of service do you use ?

**Comment by Vultour** [comment]
EMC ViPR Object Storage
